### PR TITLE
integrate the edge-tts

### PR DESCRIPTION
**Pull request by pjq**
Integrate edge-tts

### CODE REVIEW

File: xiaogpt/xiaogpt.py [hunk 1]
```diff
@@ -180,6 +191,41 @@ async def do_tts(self, value, wait_for_finish=False):
...
+
+    async def play_url(self, url):
+        url = url or f"http://{self.local_ip}:{self.port}/output.mp3"
+        print(f"play: {url}")
+        await self.mina_service.play_by_url(self.device_id, url)
+
+    async def edge_tts(self, value):
+        print(f"edge_tts")
+        self.text2mp3(value)
+        await self.play_url("")
```
Comment (frostming) [inline xiaogpt/xiaogpt.py:227]: I'd prefer this to be more explicit, the empty string doesn't make sense here. Why not return a file name by `text2mp3` and pass it to `play_url`?

**Comment by yihong0618** [comment]
I would like merge #131 first

File: xiaogpt/xiaogpt.py (hunk 1 above)
Comment (pjq) [reply xiaogpt/xiaogpt.py:227]: make sense

**Comment by pjq** [comment]
> I would like merge #131 first

Yeah, please go ahead.

File: xiaogpt/xiaogpt.py (hunk 1 above)
Comment (pjq) [reply xiaogpt/xiaogpt.py]: updated.

**Comment by yihong0618** [comment]
@pjq can you help solve the confilct?

**Comment by pjq** [comment]
@yihong0618 

I just pull the latest code, and it will throw exception
```
commit 56893f56fe7b43f37c1e6319a36741eaaa661088 (HEAD -> pjq/tmp, yihong/main)
Author: Frost Ming <me@frostming.com>
Date:   Sun Mar 19 21:21:08 2023 +0800

    feat: support stream mode and switch to async request (#131)

    * feat: support stream mode and switch to async request

    Signed-off-by: Frost Ming <me@frostming.com>

    * batch read and tts

    Signed-off-by: Frost Ming <me@frostming.com>

    ---------

    Signed-off-by: Frost Ming <me@frostming.com>
```

```
Running xiaogpt now, 用`帮我/请回答`开头来提问
或用`开始持续对话`开始持续对话
Traceback (most recent call last):
  File "/workspace/Xiaoai-ChatGPT/chatgpt-mispeaker/xiaogpt.py", line 5, in <module>
    main()
  File "/workspace/Xiaoai-ChatGPT/chatgpt-mispeaker/xiaogpt/cli.py", line 90, in main
    asyncio.run(miboy.run_forever())
  File "/opt/homebrew/Cellar/python@3.9/3.9.13_1/Frameworks/Python.framework/Versions/3.9/lib/python3.9/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/opt/homebrew/Cellar/python@3.9/3.9.13_1/Frameworks/Python.framework/Versions/3.9/lib/python3.9/asyncio/base_events.py", line 647, in run_until_complete
    return future.result()
  File "/workspace/Xiaoai-ChatGPT/chatgpt-mispeaker/xiaogpt/xiaogpt.py", line 271, in run_forever
    await self.new_record_event.wait()
  File "/opt/homebrew/Cellar/python@3.9/3.9.13_1/Frameworks/Python.framework/Versions/3.9/lib/python3.9/asyncio/locks.py", line 226, in wait
    await fut
RuntimeError: Task <Task pending name='Task-1' coro=<MiGPT.run_forever() running at /workspace/Xiaoai-ChatGPT/chatgpt-mispeaker/xiaogpt/xiaogpt.py:271> cb=[_run_until_complete_cb() at /opt/homebrew/Cellar/python@3.9/3.9.13_1/Frameworks/Python.framework/Versions/3.9/lib/python3.9/asyncio/base_events.py:184]> got Future <Future pending> attached to a different loop
```

**Comment by yihong0618** [comment]
@frostming

**Comment by yihong0618** [comment]
@pjq 
如果 tts 时间太长不太行。
----------------------------------------
self.finish_request(request, client_address)
  File "/opt/homebrew/Cellar/python@3.10/3.10.10/Frameworks/Python.framework/Versions/3.10/lib/python3.10/socketserver.py", line 360, in finish_request
    self.RequestHandlerClass(request, client_address, self)
  File "/opt/homebrew/Cellar/python@3.10/3.10.10/Frameworks/Python.framework/Versions/3.10/lib/python3.10/http/server.py", line 668, in __init__
    super().__init__(*args, **kwargs)
  File "/opt/homebrew/Cellar/python@3.10/3.10.10/Frameworks/Python.framework/Versions/3.10/lib/python3.10/socketserver.py", line 747, in __init__
    self.handle()
  File "/opt/homebrew/Cellar/python@3.10/3.10.10/Frameworks/Python.framework/Versions/3.10/lib/python3.10/http/server.py", line 433, in handle
    self.handle_one_request()
  File "/opt/homebrew/Cellar/python@3.10/3.10.10/Frameworks/Python.framework/Versions/3.10/lib/python3.10/http/server.py", line 421, in handle_one_request
    method()
  File "/opt/homebrew/Cellar/python@3.10/3.10.10/Frameworks/Python.framework/Versions/3.10/lib/python3.10/http/server.py", line 675, in do_GET
    self.copyfile(f, self.wfile)
  File "/opt/homebrew/Cellar/python@3.10/3.10.10/Frameworks/Python.framework/Versions/3.10/lib/python3.10/http/server.py", line 875, in copyfile
    shutil.copyfileobj(source, outputfile)
  File "/opt/homebrew/Cellar/python@3.10/3.10.10/Frameworks/Python.framework/Versions/3.10/lib/python3.10/shutil.py", line 198, in copyfileobj
    fdst_write(buf)
  File "/opt/homebrew/Cellar/python@3.10/3.10.10/Frameworks/Python.framework/Versions/3.10/lib/python3.10/socketserver.py", line 826, in write
    self._sock.sendall(b)

BrokenPipeError: [Errno 32] Broken pipe

**Comment by yihong0618** [comment]
@pjq 我更新了一下，你再试试？

**Comment by pjq** [comment]
好的

**Comment by pjq** [comment]
奇怪 我还是看到同样的错误

**Comment by pjq** [comment]
> ## @pjq
> 如果 tts 时间太长不太行。
> self.finish_request(request, client_address) File "/opt/homebrew/Cellar/python@3.10/3.10.10/Frameworks/Python.framework/Versions/3.10/lib/python3.10/socketserver.py", line 360, in finish_request self.RequestHandlerClass(request, client_address, self) File "/opt/homebrew/Cellar/python@3.10/3.10.10/Frameworks/Python.framework/Versions/3.10/lib/python3.10/http/server.py", line 668, in **init** super().**init**(*args, **kwargs) File "/opt/homebrew/Cellar/python@3.10/3.10.10/Frameworks/Python.framework/Versions/3.10/lib/python3.10/socketserver.py", line 747, in **init** self.handle() File "/opt/homebrew/Cellar/python@3.10/3.10.10/Frameworks/Python.framework/Versions/3.10/lib/python3.10/http/server.py", line 433, in handle self.handle_one_request() File "/opt/homebrew/Cellar/python@3.10/3.10.10/Frameworks/Python.framework/Versions/3.10/lib/python3.10/http/server.py", line 421, in handle_one_request method() File "/opt/homebrew/Cellar/python@3.10/3.10.10/Frameworks/Python.framework/Versions/3.10/lib/python3.10/http/server.py", line 675, in do_GET self.copyfile(f, self.wfile) File "/opt/homebrew/Cellar/python@3.10/3.10.10/Frameworks/Python.framework/Versions/3.10/lib/python3.10/http/server.py", line 875, in copyfile shutil.copyfileobj(source, outputfile) File "/opt/homebrew/Cellar/python@3.10/3.10.10/Frameworks/Python.framework/Versions/3.10/lib/python3.10/shutil.py", line 198, in copyfileobj fdst_write(buf) File "/opt/homebrew/Cellar/python@3.10/3.10.10/Frameworks/Python.framework/Versions/3.10/lib/python3.10/socketserver.py", line 826, in write self._sock.sendall(b)
> 
> BrokenPipeError: [Errno 32] Broken pipe

我也看到了这个错误，但好像又不影响播放

**Comment by yihong0618** [comment]
windows ?

**Comment by pjq** [comment]
> windows ?
MAC M1

**Comment by yihong0618** [comment]
> > ## @pjq
> > 如果 tts 时间太长不太行。
> > self.finish_request(request, client_address) File "/opt/homebrew/Cellar/python@3.10/3.10.10/Frameworks/Python.framework/Versions/3.10/lib/python3.10/socketserver.py", line 360, in finish_request self.RequestHandlerClass(request, client_address, self) File "/opt/homebrew/Cellar/python@3.10/3.10.10/Frameworks/Python.framework/Versions/3.10/lib/python3.10/http/server.py", line 668, in **init** super().**init**(*args, **kwargs) File "/opt/homebrew/Cellar/python@3.10/3.10.10/Frameworks/Python.framework/Versions/3.10/lib/python3.10/socketserver.py", line 747, in **init** self.handle() File "/opt/homebrew/Cellar/python@3.10/3.10.10/Frameworks/Python.framework/Versions/3.10/lib/python3.10/http/server.py", line 433, in handle self.handle_one_request() File "/opt/homebrew/Cellar/python@3.10/3.10.10/Frameworks/Python.framework/Versions/3.10/lib/python3.10/http/server.py", line 421, in handle_one_request method() File "/opt/homebrew/Cellar/python@3.10/3.10.10/Frameworks/Python.framework/Versions/3.10/lib/python3.10/http/server.py", line 675, in do_GET self.copyfile(f, self.wfile) File "/opt/homebrew/Cellar/python@3.10/3.10.10/Frameworks/Python.framework/Versions/3.10/lib/python3.10/http/server.py", line 875, in copyfile shutil.copyfileobj(source, outputfile) File "/opt/homebrew/Cellar/python@3.10/3.10.10/Frameworks/Python.framework/Versions/3.10/lib/python3.10/shutil.py", line 198, in copyfileobj fdst_write(buf) File "/opt/homebrew/Cellar/python@3.10/3.10.10/Frameworks/Python.framework/Versions/3.10/lib/python3.10/socketserver.py", line 826, in write self._sock.sendall(b)
> > BrokenPipeError: [Errno 32] Broken pipe
> 
> 我也看到了这个错误，但好像又不影响播放

是 http.server 的性能问题，未来我们得换一个。

**Comment by pjq** [comment]
感觉不是http.server的问题, 这么小的一个文件

**Comment by yihong0618** [comment]
@pjq 修复了，拉下最新代码试一下。

**Comment by yihong0618** [comment]
> 感觉不是http.server的问题, 这么小的一个文件

我研究了，不是它的问题，是小爱一次接不了那么多数据，我们得一点点传，只要 mp3 太长就会出现这个问题，你可以拿个大点的 mp3 试试就懂了。

**Comment by yihong0618** [comment]
@pjq if we succeed we can merge this

**Comment by pjq** [comment]
I test one file 1 min, 1.9M, it can play without any error.
so I am wondering some event we send to miservice cause this error.
In the past, when I using the Microsoft TTS, almost the same process, and I don't see this error.
And it seems even this error shows, it will continue to play the mp3 file.

**Comment by yihong0618** [comment]
so the task error fixed？

**Comment by yihong0618** [comment]
the same problem？
use git pull？

**Comment by pjq** [comment]
Just pull the latest code, works now.

**Comment by yihong0618** [comment]
fine let's move on

**Comment by pjq** [comment]
It seems the --config doesn't work anymore.

**Comment by yihong0618** [comment]
> It seems the --config doesn't work anymore.

just t

> It seems the --config doesn't work anymore.
> 
> The problem exists in the following code, all the config in `--config` will be over written by the options from cli params.
> 
> ```
>         if options.config:
>             config.read_from_config(options.config)
>         for key, value in vars(options).items():
>             print(f"read_from_config2 {key}={value}")
>             if value is not None and key in config.__dataclass_fields__:
>                 setattr(config, key, value)
> ```

I think its make sense cli > config

**Comment by yihong0618** [comment]
got your point....

**Comment by pjq** [comment]
We can choose the config by cli param or --config.
mixed will be hard to control.

**Comment by yihong0618** [comment]
> It seems the --config doesn't work anymore.
> 
> The problem exists in the following code, all the config in `--config` will be over written by the options from cli params.
> 
> ```
>         if options.config:
>             config.read_from_config(options.config)
>         for key, value in vars(options).items():
>             print(f"read_from_config2 {key}={value}")
>             if value is not None and key in config.__dataclass_fields__:
>                 setattr(config, key, value)
> ```

fixed

**Comment by frostming** [comment]
> all the config in `--config` will be over written by the options from cli params.

This is intentional, manually given CLI options have higher priority, this is a common practice

File: xiaogpt/cli.py [hunk 2]
```diff
@@ -53,6 +53,12 @@ def main():
         action="store_true",
         help="show info",
     )
+    parser.add_argument(
+        "--edge_tts_enable",
```
Comment (frostming) [inline xiaogpt/cli.py:57]: Let's make it `--enable_edge_tts`

File: xiaogpt/config.py [hunk 3]
```diff
@@ -68,11 +70,11 @@ def wakeup_command(self) -> str:
     @classmethod
     def from_options(cls, options: argparse.Namespace) -> Config:
         config = cls()
-        if options.config:
-            config.read_from_config(options.config)
         for key, value in vars(options).items():
             if value is not None and key in config.__dataclass_fields__:
                 setattr(config, key, value)
+        if options.config:
+            config.read_from_config(options.config)
```
Comment (frostming) [inline xiaogpt/config.py:77]: This should be fixed by #143 and this change can be reverted

File: xiaogpt/cli.py (hunk 2 above)
Comment (frostming) [reply xiaogpt/cli.py:57]: And the default value must be `None` or it will override config

**Pull request merged**